### DISCUSSION TITLE
Device coordinate problem with transparency

### DISCUSSION THREAD

**Discussion by jwwalker** [google/filament#4169] [Q&A] [answered]
I was experimenting with using an unlit material with `vertexDomain : device` to draw a background and an overlay.  It mostly works, except in how it interacts with translucent objects between the overlay and the background.  My test overlay has a device z of -1, and my test background has a device z of 0.9999.

![Screen Shot 2021-06-21 at 3 49 55 PM](https://user-images.githubusercontent.com/730426/122837495-062f4800-d2a9-11eb-8847-0abd892e9eae.png)

As you can see in this screen shot, the textured background does not show through the translucent pink ball, though other balls do show through the pink one.  And the pink one is not visible through a translucent green overlay.  This is Filament 1.10.4 tested with Metal on macOS.  Am I missing something, or is this a bug?

**Comment by romainguy** [comment, marked as answer]
Objects in the device vertex domain will still be sorted like all the other objects. You probably need to change your bounding box and/or render priority of the background and overlay.

**Reply by jwwalker** [reply]
What are the appropriate coordinates of the bounding box?  The docs on `RenderableManager::Builder::boundingBox` say it's in "object space", and I guess I assumed that if I am using NDC to specify the geometry, then I'd also use NDC for the bounding box.  I'll try the priority.

**Reply by jwwalker** [reply]
Setting priorities does fix the problem I was seeing, but I'd still like to make sure I'm doing the bounding boxes correctly.

**Reply by romainguy** [reply]
The bounding box is always in world space. But the whole point of using device coordinates is that you want the object to be visible so you might as well disable (camera) culling on it.

**Reply by jwwalker** [reply]
When I tried turning off frustum culling but not setting priorities, I still saw the problem, so apparently the bounding boxes were not the issue.  But I will turn off frustum culling as well as setting priorities, to be safe.  Thanks for your help!

**Reply by romainguy** [reply]
The object will be sorted, just not culled.

**Reply by jwwalker** [reply]
So, when you say it will be sorted, you mean I still need to worry about bounding boxes?  Or is that only if there are multiple renderables with the same priority?  I don't see how it makes sense to try to make a world-space bounding box in this situation.

**Reply by romainguy** [reply]
The bounding box will matter only if other objects have the same priority.  But the bounding box can be arbitrary, it doesn't have to have any meaning in this case.
